src/server/worker/datum/create_stream.go

```diff
@@ -0,0 +1,93 @@
+package datum
+
+import (
+	"context"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/client"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/renew"
+	"github.com/pachyderm/pachyderm/v2/src/internal/task"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+	"github.com/pachyderm/pachyderm/v2/src/pps"
+)
+
+const (
+	// The target number of files per shard. A smaller value returns datums to
+	// the client faster. A larger value reduces the overhead of creating datums.
+	ShardNumFiles = 10000
+)
+
+type createDatumStream struct {
+	ctx               context.Context
+	c                 pfs.APIClient
+	taskDoer          task.Doer
+	input             *pps.Input
+	requestDatumsChan chan bool
```

Comment:
These should be `chan<- bool`, I think.  It's convenient to initialize this struct with `make(chan bool)` and then probe the privates to read from the channel, but it's poor documentation for how the channel is actually used.  The caller should initialize a channel and methods on this struct should only be able to send.

Reply:
Can you elaborate more? The struct does both. In `createPFS()`, we're reading from this channel and in `Iterate()` we're writing to this channel.

Reply:
So in createDatumStream you only recv from requestDatumsChan, and you only send to fsidChan and errChan.  Then below in createDatumStreamIterator, you send to requestDatumsChat and recv from fsidChan and errChan.  The types should correspond to that.

Also, I don't think you're using the boolean value in requestDatumsChan.  In that case, use `chan struct{}`.  (`requestDatumsChan <- false` is the same as `requestDatumsChan` <- true, so just do `requestDatumsChan <- struct{}{}`).

The way to set this up is like:

```go
type X
type sender struct {
   ch chan<- X
}
type receiver struct{
   ch <-chan X
}
func whatever() {
    ch := make(chan X)
    s := &sender{ch: ch}
    r := &receiver{ch: ch}
    s.ch <- x // ok
    x = <- r.ch // ok
    x = <- s.ch // compile error
    r.ch <- x // compile error
}
```
I think it's important to do this because the channels are used in two separate files.  Making sure you only send in one and receive in the other is something the compiler can validate for you, and is good design.  If you accidentally start receiving in the sender, your code will just be confusingly buggy.  The compiler can save you.

Reply:
I see what you mean. 

I define `createDatumStreamIterator` with an embedded `createDatumStream` which means I access the channels via the embedded type. How can I achieve this type definition? Would I need to create a duplicate `createDatumStream` just to have one struct definition specify one set of channel directions and the duplicate struct to specify the other set of directions?

Reply:
I see that, but why is it an embedded struct?  If the iterator only needs 3 channels, I would just pass the 3 channels to the iterator.  I'm actually not even sure why it's two structs; you can't create either the stream or the stream iterator from outside this package; you can only call the public constructor which creates both at the same time.

One thing that tests don't cover is the case where the stream sends an error to the error channel.  If your iterator only takes the channels, then you can write a test like:

```go
func TestIteratorError(t *testing.T) {
   errCh := make(chan error)
   it := createDatumStreamIterator{ errCh: errCh }
   go func(} { errCh <- errors.New("error") }
   _, err := it.Iterate()
   require.Error(err)
```